crypto/evp/keymgmt_lib.c

```diff
@@ -0,0 +1,307 @@
+
+        /* If the legacy key doesn't have an export function, give up */
+        if (pk->ameth->export_to == NULL)
+            return NULL;
+
+        /* Otherwise, simply use it */
+        provkey = pk->ameth->export_to(pk, keymgmt);
+    } else {
+        /*
+         * Here, there is no legacy key, so we look at the already
+         * cached but there are cached provider keys, and import from
+         * the first that supports it (i.e. use its export function),
+         * and export the imported data to the new provider.
+         */
+
+        /*
+         * If the given keymgmt doesn't have an import function, give up
+         */
+        if (keymgmt->importkey_priv == NULL)
+            return NULL;
+
+        for (j = 0; j < i && pk->pkeys[j].keymgmt != NULL; j++) {
+            if (pk->pkeys[j].keymgmt->exportkey_priv != NULL) {
+                const OSSL_PARAM *paramdefs = NULL;
+                OSSL_PARAM *params = NULL;
+                void *data = NULL;
+                void *provctx =
+                    ossl_provider_ctx(EVP_KEYMGMT_provider(keymgmt));
+
+                paramdefs = pk->pkeys[j].keymgmt->exportkey_priv_types();
+                /*
+                 * All params have 'data' set to NULL.  In that case,
+                 * the exportkey call should just fill in 'return_size'
+                 * in all applicable params.
+                 */
+                params = paramdefs_to_params(paramdefs);
+                /* Get 'return_size' filled */
+                pk->pkeys[j].keymgmt->exportkey_priv(pk->pkeys[j].provkey,
+                                                     params);
+
+                /*
+                 * Allocate space and assign 'data' to point into the
+                 * data block
+                 */
+                data = allocate_params_space(params);
+
+                /*
+                 * Call the exportkey function a second time, to get
+                 * the data filled
+                 */
+                pk->pkeys[j].keymgmt->exportkey_priv(pk->pkeys[j].provkey,
+                                                     params);
+
+                /*
+                 * We should have all the data at this point, so import
+                 * into the new provider and hope to get a key back.
+                 */
+                provkey = keymgmt->importkey_priv(provctx, params);
+                OPENSSL_free(params);
+                OPENSSL_free(data);
+
+                if (provkey != NULL) {
+                    break;
+                }
+            }
+        }
+    }
+
+    /*
+     * TODO(3.0) Right now, we assume we have ample space.  We will
+     * have to think about a cache aging scheme, though, if |i| indexes
+     * outside the array.
+     */
+    j = ossl_assert(i < OSSL_NELEM(pk->pkeys));
+
+    EVP_KEYMGMT_up_ref(keymgmt);
+    pk->pkeys[i].keymgmt = keymgmt;
+    pk->pkeys[i].provkey = provkey;
```

Comment:
Shouldn't there be a test for provkey being NULL somewhere around here?

Reply:
Why?  What will we do as a result...  apart from returning NULL?

Reply:
Well we wouldn't up ref keymgmt or cache it in that case.

Reply:
Ooooh, you're right!  Thanks, eagle-eye.  Will fix.

Reply:
Done